### PR TITLE
✨ GUI - Implementation of the Zappy's objects

### PR DESCRIPTION
**Pull request by BxptisteM**
- Added main and args management
- Added an IException interface and added abstract AException
- Added interface IObject
- Added abstract classes for the objects and the stones (also inherits from objects)
- Added all the stones
- Added food
- Added eggs
- Added 37 tests on the objects

### CODE REVIEW

File: gui/src/handle_args/HandleArgs.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include <cstdlib>
+#include <iostream>
+
+using namespace GUI;
+
+int HandleArgs::checkArgs(int nbArgs, char **args)
+{
+    try {
+        if (nbArgs == 2 &&
+            (std::string(args[1]) == "-help" || std::string(args[1]) == "-h")) {
```
Comment (RenardFute) [inline gui/src/handle_args/HandleArgs.cpp:22]: It's an error in the subject, I saw it with Hugo, and it shall have two '-' which result in '--help' and '-h' optionaly

File: gui/src/handle_args/HandleArgs.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            printUsage();
+            goto end;
```
Comment (RenardFute) [inline gui/src/handle_args/HandleArgs.cpp:24]: falling

Pleases remove goto, you can return here directly if you want to do so

File: gui/src/main.cpp [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
+using namespace GUI;
 
 int main(int ac, char **av)
 {
-    (void)ac;
-    (void)av;
-    hello_world();
+    GUI::HandleArgs args;
+
+    args.checkArgs(ac, av);
```
Comment (RenardFute) [inline gui/src/main.cpp:17]: GUI::HandleArgs#checkArgs gives you a return code, you should exit if it's > 0, no ?

File: gui/src/objects/abstracts/AStone.hpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#ifndef ASTONE_HPP_
+#define ASTONE_HPP_
+
+#include "AObject.hpp"
+
+namespace GUI {
+class AStone : public AObject {
+    public:
+        /**
+         * @brief Construct a new AStone object
```
Comment (RenardFute) [inline gui/src/objects/abstracts/AStone.hpp:17]: Nice documentation but it's missing from many other classes

File: tests/gui/get_objects_density_spec.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    ASSERT_FLOAT_EQ(0.05, thystame.getDensity());
+}
+
+TEST(get_objects_density, test7)
+{
+    GUI::Food food;
+    ASSERT_FLOAT_EQ(0.5, food.getDensity());
+}
+
+// NOW TESTS WITH UNIQUE PTS OBJECTS //
```
Comment (RenardFute) [inline tests/gui/get_objects_density_spec.cpp:60]: 👀 A little help for your copilot I see x)

**Review by Quentintnrl — APPROVED** [review]

**Review by RenardFute — CHANGES_REQUESTED** [review]
The architecture in the overall seems fine to me.
I saw you started to make a bit of documentation but didn't go all the way, can you finish that ?
Also, just little warnings, your indentation in your headers is not the preferred one, may you fix this please.
Second little heads up, if you store a value (e.g. in a constructor) you shall pass it as copy and then move it using ``std::move``, references are good when you are doing operations on a variable, and you don't want to copy it over and over but when you to "pass"/"give" a value like in setters or constructor ``std::move`` is better.

I know you are using VSCode so you don't have such warnings, but in front of your methods/functions declaration (in header files), if they return something which is not void, you may want to add ``[[nodiscard]]``. It's a tag that tells others that the returned value shall not be ignored.

I know the review is starting to be quite big, but not everything is mandatory, if you can, in your documentation when you add ``@return`` or ``@param`` give a little more than just the type or the name it would be great ^^

File: gui/src/main.cpp (hunk 3 above)
Comment (BxptisteM) [reply gui/src/main.cpp:17]: It's my bad i already throw exceptions in the checkArgs function, im going to change it to return void

File: gui/src/exception/ArgsException.cpp [hunk 6]
```diff
@@ -10,6 +8,6 @@
 using namespace GUI;
 
 ArgsException::ArgsException(const std::string &message)
```
Comment (RenardFute) [inline gui/src/exception/ArgsException.cpp:10]: Just fyi, what I meant was actually this:
```suggestion
ArgsException::ArgsException(std::string message): AException(std::move(message), "ArgsException")
```

But it's ok, just you have time fix this ^^

File: gui/src/exception/abstract/AException.hpp [hunk 7]
```diff
@@ -1,30 +1,43 @@
...
+        */
         virtual ~AException() noexcept = default;
 
-        const char *what() const noexcept override final;
-        std::string getType() const noexcept override final;
+        /**
+         * @brief Get the message object
+         * @return const char* (error message)
+        */
+        [[nodiscard]] const char *what() const noexcept override final;
```
Comment (RenardFute) [inline gui/src/exception/abstract/AException.hpp:30]: If you have ``final`` ``override`` is induced so you can remove ``override``

File: gui/src/exception/interface/IException.hpp [hunk 8]
```diff
@@ -1,24 +1,36 @@
...
         virtual ~IException() noexcept = default;
 
-        virtual const char* what() const noexcept override = 0;
-        virtual std::string getType() const noexcept = 0;
-};
+        /**
+         * @brief Get the what object
+         * @return const char* (error message)
+        */
+        [[nodiscard]] virtual const char* what() const noexcept override = 0;
```
Comment (RenardFute) [inline gui/src/exception/interface/IException.hpp:28]: ``override`` induces ``virtual`` so you can remove it.

File: gui/src/objects/food/Food.cpp [hunk 9]
```diff
@@ -1,6 +1,4 @@
 /*
-** EPITECH PROJECT, 2024
```
Comment (RenardFute) [inline gui/src/objects/food/Food.cpp]: Sorry I just meant, removing 'WSL ...' the Epitech tag and project name shall stay

File: tests/gui/get_objects_density_spec.cpp [hunk 10]
```diff
@@ -57,8 +55,6 @@ TEST(get_objects_density, test7)
     ASSERT_FLOAT_EQ(0.5, food.getDensity());
 }
 
-// NOW TESTS WITH UNIQUE PTS OBJECTS //
```
Comment (RenardFute) [inline tests/gui/get_objects_density_spec.cpp]: You could have let it there, it just made me laugh ^^

**Review by RenardFute — CHANGES_REQUESTED** [review]
I'll approve this, so you don't have to re-request me but, may you fix the headers like this:
```C++
/*
** EPITECH PROJECT, 2024
** Zappy
** File description:
** .....
*/
```
And then merge. Thanks

**Review by 6im0n — COMMENTED** [review]
👍 good for me !

**Review by RenardFute — APPROVED** [review]

**Pull request merged**